core/payment/src/dao/allocation.rs

```diff
@@ -152,12 +152,31 @@ impl<'c> AllocationDao<'c> {
         .await
     }
 
-    pub async fn release(&self, allocation_id: String, owner_id: Option<NodeId>) -> DbResult<bool> {
+    pub async fn release(
+        &self,
+        allocation_id: String,
+        owner_id: Option<NodeId>,
+    ) -> DbResult<AllocationReleaseStatus> {
         let id = allocation_id.clone();
         do_with_transaction(self.pool, move |conn| {
+            let mut query = dsl::pay_allocation
```

Comment:
I think this code could be much simpler. First you can get allocation without filtering, than check conditions (owner_id and if allocation is released). Than you can update released field, knowing that it will succeed